2단계 - 문자열 계산기

giwankim opened the pull request:
역할 분할을 생각하면서 구현해봤습니다.

- IoC 컨테이너가 없어서 Calculator 클래스는 통합 테스트로 테스트하게 되었습니다. 개발할 떼 인터페이스 분리하고 DI에 의존해왔는데 프레임워크 없이 구현하는 것도 오랜만이네요. ㅋㅋㅋ
- 발리테이션 로직을 여기저기 겍체애 흝어지게 구현했습니다.
- 확장성을 고려해서 파싱 기능을 구분했습니다. 컴파일러를 공부해본 적은 없는데 아마 이렇게 Parser가 Tree 구조를 만들고 트리를 값으로 evaluate하는 대략적인 구조라고 들은 것 같습니다.

자바는 사용해봤는데 코틀린 사용이 처음이라 코틀린스러운 문법을 사용했는지 모르겠습니다.

우선은 익숙한 Junit5를 사용했는데 마침 https://techblog.woowahan.com/5825/ 에 비슷한 예제로 코테스트 활용이 있어서 읽어볼 생각입니다.

몇일 코로나로 🤮  고생했는데 몸조심하세요...

giwankim committed:
test: Kotest를 사용해서 1단계 미션과 동일한 테스트를 추가했다.

giwankim committed:
test: 사용하게 될 String 함수와 코들린 확장 함수를의 학습 테스트를 진행했다.

giwankim committed:
feat: 사칙언산 enum 클래스를 구현했다.

각 사칙 연산은 두 개의 operand 숫자로 계산할 수 있는 메소드들을 override 하고 있다.

giwankim committed:
refactor: fromSymbol 로직을 단순화

giwankim committed:
feat: 파싱된 결과물인 Expr 인터페이스와 구현체를 추가했다.

giwankim committed:
feat: 파싱의 대상이 operator 들과 operand 들의 1급 컬렉션을 추가했다.

giwankim committed:
feat: 일력값을 파싱하는 클래스인 Parser 를 구현했다.

giwankim committed:
feat: 계산기 구현과 통합 테스트를 추가했다.

Hyeon9mak commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,9 @@
+package calculator
+
+class Calculator {
+    fun calculate(input: String): Int {
+        require(input.isNotBlank()) { "입력값이 빈 공백 문자입니다." }
+        val expression: Expr = Parser(input).parse()
+        return expression.evaluate()
+    }
```
require 를 이용한 IllegalArumentException 검증 👍

Hyeon9mak commented on src/main/kotlin/calculator/Expr.kt:
```diff
@@ -0,0 +1,19 @@
+package calculator
+
+sealed interface Expr
+
+data class Literal(
+    val value: Int,
+) : Expr
+
```
`sealed` 로 범위를 한정시켜주셨군요 👍 

data class 와 value class 의 차이는 무엇일까요?
둘의 차이를 확인하고 조금 더 적합하다 싶은 구조를 선택해줘도 좋을거 같아요.

giwankim replied:
이 경우 Literal을 value class로 선언하면 불필요한 클래스 할당을 줄일 수 있겠군요.

사실 본질적으로 해결하려던 문제는 테스트를 위해 equals 메서드가 필요했다는 점입니다. 파싱 결과물을 비교하는 ParserTest에서 필요하게 되었는데요, Java로 구현했다면 별 고민 없이 equals와 hashCode를 오버라이드했을 텐데, Kotlin 컴파일러의 자동 생성 기능을 최대한 활용하는 방향으로 가닥을 잡았습니다.

코틀린 클래스에서도 equals, hashCode, toString 메서드를 자주 오버라이드하게 되나요? 그리고 테스트를 위해 객체 간의 equals 메서드를 구현하는 것이 찜찜한 면도 없지 않아 있습니다.

Hyeon9mak commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,40 @@
+package calculator
+
+enum class Operator(
+    val symbol: String,
+) {
+    ADD("+") {
+        override fun apply(
+            x: Int,
+            y: Int,
+        ): Int = x + y
+    },
+    SUBTRACT("-") {
+        override fun apply(
+            x: Int,
+            y: Int,
+        ): Int = x - y
+    },
+    MULTIPLY("*") {
+        override fun apply(
+            x: Int,
+            y: Int,
+        ): Int = x * y
+    },
+    DIVIDE("/") {
+        override fun apply(
+            x: Int,
+            y: Int,
+        ): Int = if (y != 0) x / y else throw IllegalArgumentException("0으로 나눌 수 없습니다.")
+    }, ;
+
+    abstract fun apply(
+        x: Int,
+        y: Int,
+    ): Int
```
👍 👍 
람다 표현식을 활용해서 요렇게 줄여볼 수도 있겠네요 😉 

```kotlin
enum class Operator(
    val symbol: String,
    val operation: (Int, Int) -> Int
) {
    ADD("+", { x, y -> x + y }),
    ;
}
```

Hyeon9mak commented on src/main/kotlin/calculator/Expr.kt:
```diff
@@ -0,0 +1,19 @@
+    val value: Int,
+) : Expr
+
+data class Binary(
+    val left: Expr,
+    val operator: Operator,
+    val right: Expr,
+) : Expr
+
+fun Expr.evaluate(): Int =
+    when (this) {
+        is Literal -> this.value
+        is Binary -> this.operator.apply(left.evaluate(), right.evaluate())
+    }
```
연산자면 좌우를 타고 내려가면서 재귀를 돌고, 아니면 그 값을 취한다~

Hyeon9mak commented on src/main/kotlin/calculator/Parser.kt:
```diff
@@ -0,0 +1,32 @@
+        private val DELIMITER_REGEX = "\\s+".toRegex()
+    }
+}
+
+private fun Int.isOdd(): Boolean = this % 2 == 1
```
클래스 내부에 위치시켜도 괜찮을거 같은데요? 😉

Hyeon9mak commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,62 @@
+class CalculatorTest {
+    private val calculator = Calculator()
+
+    @ParameterizedTest
+    @CsvSource(
+        value = [
+            "1 + 2 : 3",
+            "2 - 1 : 1",
+            "2 * 3 : 6",
+            "4 / 2 : 2",
+            "1 + 2 * 3 : 9",
+            "2 + 3 * 4 / 2 : 10",
+        ],
+        delimiterString = " : ",
+    )
+    fun `사칙연산을 수행한다`(
+        input: String,
+        expected: Int,
+    ) {
```
테스트 가독성이 엄청 뛰어나네요 👍

Hyeon9mak commented on src/test/kotlin/calculator/ExprTest.kt:
```diff
@@ -0,0 +1,51 @@
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+
+@Suppress("NonAsciiCharacters")
+class ExprTest {
+    @Test
+    fun add() {
+        val expr = Binary(Literal(1), Operator.ADD, Literal(2))
+        val result = expr.evaluate()
+        assertThat(result).isEqualTo(3)
+    }
+
+    @Test
```
테스트를 꼼꼼히 잘 작성, 활용해주시는거 같아요!
다음 단계에서는 [kotest](https://kotest.io/) 를 활용해보셔도 좋을거 같네요

Hyeon9mak commented on src/test/kotlin/study/PersonKoTest.kt:
```diff
@@ -0,0 +1,31 @@
+package study
+
+import io.kotest.core.spec.style.StringSpec
```
이전 단계 코드도 꼼꼼히 패키지 관리 👍

Hyeon9mak reviewed (changes requested):
안녕하세요 지완님! 2단계도 잘 진행해주셨습니다 👍 👍 
여러가지 고민을 많이해주신 흔적들이 보입니다.
아주 간단한 코멘트만 몇 가지 남겼어요!

IoC 컨테이너 와 통합 테스트 관련 코멘트는 제가 잘 이해를 못했는데요, 정확히 어떤 테스트를 하고 싶으셨던걸까요? 자세한 설명을 주신다면 최대한 도움 드릴 수 있을거 같습니다! 🙇‍♂️ 

궁금하신 점 얼마든지 질문 남겨주세요~

giwankim committed:
refactor: 연산자 enum 클래스를 람다 표현식을 생성자 인자로 활용하여 단순화 하였다.

giwankim committed:
refactor: 홀수인지 체크하는 메서드를 클래스 내부로 옮겼다.

아직은 홀수인지 체크하는 메서드가 한군데에서만 사용하고 있으니 굳이 Int 의 extension function 을 사용할 필요가 없다고 판단된다.

giwankim committed:
refactor: inline class 를 사용해서 성능을 개선했다.

Hyeon9mak commented on src/main/kotlin/calculator/Expr.kt:
```diff
@@ -0,0 +1,20 @@
+
+sealed interface Expr
+
+@JvmInline
+value class Literal(
+    val value: Int,
+) : Expr
+
```
> 코틀린 클래스에서도 equals, hashCode, toString 메서드를 자주 오버라이드하게 되나요? 그리고 테스트를 위해 객체 간의 equals 메서드를 구현하는 것이 찜찜한 면도 없지 않아 있습니다.

잘 아시겠지만 equals, hashCode, toString 이 제공하는 기능이 필요하다면 얼마든지 오버라이드 하게 되죠. 결국 메서드 중 하나일 뿐이니까요.

data class, value class 는 '알아서 오버라이드 해준다.' 외에도 '다른 사람이 읽을 때 어떤 용도로 만들어진 클래스인지 힌트를 제공해준다.' 라는 의미도 있는거 같아요.

Hyeon9mak commented on src/main/kotlin/calculator/Operator.kt:
```diff
@@ -0,0 +1,21 @@
+package calculator
+
+enum class Operator(
+    val symbol: String,
+    val operation: (Int, Int) -> Int,
+) {
+    ADD("+", { x, y -> x + y }),
+    SUBTRACT("-", { x, y -> x - y }),
+    MULTIPLY("*", { x, y -> x * y }),
+    DIVIDE("/", { x, y -> if (y != 0) x / y else throw IllegalArgumentException("0으로 나눌 수 없습니다.") }), ;
+
+    fun apply(
+        x: Int,
+        y: Int,
+    ): Int = operation(x, y)
+
```
`operation` 이 public 하게 공개되었으므로, `apply` 메서드를 제거해볼 수도 있겠네요 😉

Hyeon9mak reviewed (approved):
지완님 피드백 잘 반영해주셨습니다.
질문 주신 내용 답변 코멘트 남겼으니 확인해주세요!
다음 단계에서 뵙겠습니다~
